rand_distr/src/dirichlet.rs

```diff
@@ -58,8 +58,8 @@ where StandardNormal: Distribution<N>, Exp1: Distribution<N>, Open01: Distributi
         if a.len() < 2 {
             return Err(Error::AlphaTooShort);
         }
-        for i in 0..a.len() {
-            if !(a[i] > N::from(0.0)) {
+        for &ai in &a {
+            if !(ai > N::from(0.0)) {
```

Comment:
This is one of those style things where I simply don't agree with Clippy that we should avoid indices.

Also, we have a `N: Copy` restriction which implies there's no need for `&ai` here (also below).

Reply:
I think in general it reduces the number of bound checks the compiler has to perform, so it is not just a cosmetic change. However, I'm not sure it makes a difference here.

Reply:
Does it really? Those bounds-checks should be very easy to eliminate.

Reply:
In general, the bound checks cannot be eliminated, because it would change when the program panics. Consider the following program:

```rust
fn f(v: &[i32]) {
    for i in 0..10 {
        println!("{}", v[i]);
    }
}
```

This can be fixed by avoiding to use indices or by adding `assert!(v.len() >= 10);` before the loop.

Reply:
We're not talking about in general.

Reply:
I think the general case matters for consistency, but I don't feel strongly about this. Do you want me to revert this change?

> Also, we have a N: Copy restriction which implies there's no need for &ai here (also below).

I don't think this is possible, only `into_iter` allows iterating `f64` instead of `&f64`, but it consumes the vector.

Reply:
No, lets talk only about the specific case. Our standard review procedure is to check every change (or otherwise satisfy oneself that all are acceptable). I don't think we should have to change this procedure because of Clippy.

> I don't think this is possible, only into_iter allows iterating f64 instead of &f64

Then use `.map(|&x| x)`. This kind of explicitness is the disadvantage of iterators. Otherwise you are only trading reliance on the compiler to eliminate unneeded bounds checks with reliance on the compiler to eliminate unneeded references.

Reply:
> Then use .map(|&x| x).

This is equivalent to what I'm currently doing: `for &ai in &a` vs. `for ai in a.iter().map(|&x| x)` Do you think the second variant is preferable? The first one feels more idomatic to me.

> Otherwise you are only trading reliance on the compiler to eliminate unneeded bounds checks with reliance on the compiler to eliminate unneeded references.

I think that is a good trade, because the references can be eliminated more easily than the bound checks. As demonstrated above, the latter is sometimes not even a valid optimization the compiler can make.

Reply:
So is it equivalent or is it a trade? I'm not familiar enough with match semantics to answer that.

As far as I can see, the demonstration above is irrelevant to this case since the bounds are known.

Really, I see no good reason to accept this change, but also no good reason against, so lets just go with it.